src/action/network.rs

```diff
@@ -0,0 +1,477 @@
+    // Optimization: when we are asked only for listening connections,
+    // we need only TCP protocol.
+    let protocol_flags = if request.listening_only {
+        netstat2::ProtocolFlags::TCP
+    } else {
+        netstat2::ProtocolFlags::all()
+    };
+
+    let connection_iter = netstat2::iterate_sockets_info(
+        addr_family_flags,
+        protocol_flags
+    );
+    let connection_iter = match connection_iter {
+        Ok(val) => val,
+        Err(err) => return Err(Error::from(err).into()),
+    };
+
+    for connection in connection_iter {
+        let connection = match connection {
+            Ok(val) => val,
+            Err(err) => {
+                error!("unable to get socket information: {}", err);
+                continue;
+            },
+        };
+        let socket_info = connection.protocol_socket_info;
+
+        if request.listening_only {
+            use ProtocolSocketInfo::{Tcp, Udp};
+
+            let tcp_info = match &socket_info {
+                Tcp(tcp_info) => tcp_info,
+                Udp(_) => {
+                    // We are interested only in TCP connections.
+                    continue;
+                },
+            };
+            if tcp_info.state != TcpState::Listen {
+                // We are interested only in listening connections.
+                continue;
+            }
+        }
+
+        let pids = &connection.associated_pids;
+
+        if pids.is_empty() {
+            // If the process ID is not associated, we send a response
+            // without it. The behavior is different from Python implementation,
+            // because the latter lists only the connections associated with a
+            // process.
+            session.reply(Response {
+                socket_info: socket_info,
+                process_info: None,
+            })?;
+            continue;
+        }
+
+        // Otherwise, we have one or more processes associated with the
+        // connection. Since the protocol allows to specify only one process
+        // for connection, we report this connection multiple times, as if the
+        // associated processes had different connections. This behavior is
+        // also compatible with the Python implementation, which iterates over
+        // all the processes and then iterates over the connections for each
+        // of them.
+        for pid in pids {
+            session.reply(Response {
+                socket_info: socket_info.clone(),
+                process_info: Some(ProcessInfo::from_system(&system, *pid)),
+            })?;
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use netstat2::TcpSocketInfo;
+    use std::net::{TcpStream, TcpListener, UdpSocket, SocketAddr};
+
+    /// Returns all the responses whose `local_addr` is equal to `addr`.
+    fn find_responses_by_addr<'a>(
+        session: &'a session::test::Fake,
+        addr: SocketAddr,
+    ) -> Vec<&'a Response> {
```

Comment:
It seems like this explicit lifetime specifier is redundant here. Initially each parameter that is a reference gets its own lifetime parameter. And then compiler applies second lifetime elision rule here. 
Quote from the book: "The second rule is if there is exactly one input lifetime parameter, that lifetime is assigned to all output lifetime parameters: `fn foo<'a>(x: &'a i32) -> &'a i32.`"

Reply:
Yes, you are right :) Thanks, it's fixed now.